core/src/main/java/org/apache/carbondata/core/scan/model/QueryModel.java

```diff
@@ -209,13 +215,40 @@ private static void setDimAndMsrColumnNode(FilterProcessVO processVO, ColumnExpr
       if (null != isFilterMeasures) {
         isFilterMeasures[msr.getOrdinal()] = true;
       }
-    } else {
+    } else if (null != CarbonUtil.findDimension(processVO.getImplicitDimensions(), columnName)) {
       // check if this is an implicit dimension
-      dim = CarbonUtil
-          .findDimension(processVO.getImplicitDimensions(), columnName);
+      dim = CarbonUtil.findDimension(processVO.getImplicitDimensions(), columnName);
       col.setCarbonColumn(dim);
       col.setDimension(dim);
       col.setDimension(true);
+    } else {
+      // in case of sdk or fileformat, there can be chance that each carbondata file may have
+      // different schema, so every segment properties will have dims and measures based on
+      // corresponding segment. So the filter column may not be present in it. so generate the
+      // dimension and measure based on the info present in column expression
+      col.setDimension(isDimensionFromExpression);
+      col.setMeasure(isMeasureFromExpression);
+      if (isDimensionFromExpression) {
+        col.setDimension(new CarbonDimension(col.getCarbonColumn().getColumnSchema(),
```

Comment:
Column expression will have the CarbonColumn, but in this method, first we do reset, which set the dimension and measure and boolean values to null, so if column is not present it will not able to get it again, so im constructing the dimension or measure based on boolean flag in expression from CarbonColumn

Reply:
where it sets the `CarbonColumn` in `ColumnExpression`?

Reply:
yes, you are correct, handled